homeassistant/components/romy/config_flow.py

```diff
@@ -0,0 +1,106 @@
+"""Config flow for ROMY integration."""
+from __future__ import annotations
+
+from typing import Any, Optional
+
+import romy
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import zeroconf
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD
+from homeassistant.data_entry_flow import FlowResult
+import homeassistant.helpers.config_validation as cv
+
+from .const import DOMAIN, LOGGER
+
+
+def _schema_with_defaults(
+    default_values: Optional[dict[str, Any]] = None,
+    requires_password: bool = False,
+) -> vol.Schema:
+    if default_values is None:
+        default_values = {}
+    schema = {
+        vol.Required(CONF_HOST, default=default_values.get(CONF_HOST, "")): cv.string,
+        vol.Optional(CONF_NAME, default=default_values.get(CONF_NAME, "")): cv.string,
+    }
+
+    if requires_password:
+        schema[vol.Required(CONF_PASSWORD)] = vol.All(str, vol.Length(8))
+
+    return vol.Schema(schema)
+
+
+class RomyConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle config flow for ROMY."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self) -> None:
+        """Handle a config flow for ROMY."""
+        self.discovery_schema = None
+
+    async def async_step_user(
+        self, user_input: dict[str, str] | None = None
+    ) -> FlowResult:
+        """Handle the user step."""
+        errors: dict[str, str] = {}
+        data = self.discovery_schema or _schema_with_defaults()
+
+        if user_input is not None:
+            # Save the user input and finish the setup
+            new_romy = await romy.create_romy(
+                user_input[CONF_HOST], user_input.get(CONF_PASSWORD, "")
+            )
+
+            # get robots name in case none was provided
+            if not user_input[CONF_NAME]:
+                user_input["name"] = new_romy.name
+
+            if not new_romy.is_initialized:
+                errors[CONF_HOST] = "cannot_connect"
+
+            if not new_romy.is_unlocked:
+                data = _schema_with_defaults(user_input, requires_password=True)
+                errors[CONF_PASSWORD] = "invalid_auth"
+
+            if not errors:
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME], data=user_input
+                )
+
+        return self.async_show_form(step_id="user", data_schema=data, errors=errors)
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+
+        LOGGER.debug("Zeroconf discovery_info: %s", discovery_info)
+
+        # extract unique id and stop discovery if robot is already added
+        unique_id = discovery_info.hostname.split(".")[0]
```

Comment:
romy robots use the uniqueID as hostname which is derived from the allwinner H3 based SoC hardware uniqueID
so its hopefully very unlikely the hostname gets changed (except maybe your robot get hacked)

Reply:
Can you add this context as a comment? I think this is very important context here as otherwise it looks incorrect to do this.

What unique id will we use when the integration is configured manually? This seems like it won't work if the user configures the robot manually and it is also then discovered via zero confg

I saw the device has a `unique_id` field, and then we're using that for the actual device unique id -- does it make sense to use that here too for the config entry unique id?

Reply:
you are right its quite confusing, removed getting uniqueID from hostname,
getting it now unified from romy api